od: implement remaining functionality

wimh opened the pull request:
It does not take very long any more until I finish the remaining functionality of `od`. 
I'll have some questions, but there will probably also be remarks on what I have done so far.

todo:
- [x] fix command-line parsing of file names (eg. in `od -A x -- -f`, `x` is not a file name, but `-f` is)
- [x] ~~implement `--strings` (?)~~
- [x] implement `+size` to skip bytes
- [x] implement `--traditional`
- [x] add support for 16-bit floats
- [x] ~~support for long double floats (`f128`) (?)~~
- [x] ~~support for `i128`/`u128` (?)~~
- [x] improve/extend `--help` text
- [x] refactor odfunc/print_bytes to limit the number of parameters
- [x] refactor huge functions in `od.rs`, `parse_formats.rs` and `MultiFileReader.rs` to smaller functions.
- [x] update `od` status in `README.md`

See also the manpages at https://www.gnu.org/software/coreutils/manual/html_node/od-invocation.html#od-invocation and http://www.unix.com/man-page/posix/1p/od/

this closes #237

wimh commented:
Some remarks/questions I have:

- `--strings` prints null terminated strings in a binary input. A very basic version of the `strings` utility from GNU Binutils. Is this still useful or it it obsolete? (And if I would implement is, should it support ASCII only or utf-8)?
- `f128` is not natively supported by rust (any more). I did not find a crate which does either...
- while searching on crates.io, I did find a [f16 implementation](https://crates.io/crates/half), so I probably can implement that. 
- `i128`: native 128-bit integer support will be implemented (RFC 1504), so is it better to wait or use the [extprim crate](https://crates.io/crates/extprim) now? 
- Error messages are not always very specific. For example `od --skip-bytes=1YB` is valid syntax, but a yotta-byte does not fit into a usize. The current error message is invalid argument...
- I probably can move some parts into uucore, but I think it is better to use a separate pull request for that to avoid merge conflicts/rebasing.
- Sizes like `S`, `I`, `L` refer to the c compiler size of short, int and long int. I hard-coded this to 16-, 32- and 64-bit.
- `MultifileReader` looks similar to the [fileinput crate](https://crates.io/crates/fileinput). I want to check if merge my changes and use that. 
- I want to check if I can test this on a raspberry pi, as it is big endian.

nathanross commented:
Wow, this is amazing!! It may be a little while before other contributors have the time to really dig into this PR. If you'd like immediate progress you may break it up a bit, but feel free to leave it as is as well.

wimh commented:
I don't need immediate progress, because there is still enough to do for me. And I understand that reviewing takes time, that why I created this: to at least allow others giving me feedback while I am still working on it. For example, if you don't like the first commit, I have to change almost every commit afterwards because all new integration tests are created the same way. 
But if someone does not want to review until I say it is finished, that's fine too.

But I can split it into separate PRs if that's easier, this PR will contain just the remaining commits after a rebase. For example the first 6 commits look like a good start to me. Just let me know.

nathanross commented on src/od/od.rs:
```diff
@@ -74,418 +96,336 @@ pub fn uumain(args: Vec<String>) -> i32 {
+fn odfunc(line_bytes: usize, input_offset_base: Radix, byte_order: ByteOrder,
+        fnames: Vec<InputSource>, formats: &[ParsedFormatterItemInfo], output_duplicates: bool,
+        skip_bytes: usize, read_bytes: Option<usize>) -> i32 {
+
```
Because there are now many function inputs in this complex function, it may be good to organize the inputs into a struct, to make it is easy to differentiate within the function between function inputs and derivations based on those.

wimh replied:
I agree, 8 parameters is clearly a code smell. (the same applies to print_bytes). 

Besides that, the formats parameter contains the  ParsedFormatterItemInfo struct, which wraps a FormatterItemInfo struct and that wraps a FormatWriter struct. In odfunc, ParsedFormatterItemInfo is wrapped into a SpacedFormatterItemInfo. Probably a bit too much layering which should be decoupled too.

nathanross commented on src/od/od.rs:
```diff
@@ -74,418 +96,336 @@ pub fn uumain(args: Vec<String>) -> i32 {
+                // not enough byte for a whole element, this should only happen on the last line.
+                if n != line_bytes {
+                    // set zero bytes in the part of the buffer that will be used, but is not filled.
+                    let mut max_used = n + MAX_BYTES_PER_UNIT;
```
In ```od.rs```, ```parse_formats.rs``` and to a lesser extent ```MultiFileReader.rs``` there are currently some lengthy functions that should be at some point broken up into separate functions so as to make it easier to understand at a glance purpose different sections of their code do / should serve. 

 (e.g. this code could be moved into a small function called ```zero_remaining_buffer_bytes()``` or something like that).

wimh commented:
I want to finalize this PR now, and not implement any new features. Of course I am still going to fix issues related to this PR.

I think it is better not to implement `i128`, `u128` and `f128` support until rust natively supports it.

I also did not implement `--strings` at this moment. It would make more sense if `strings` from gnu binutils was ported to rust, so it could share code, or `od` could simply call `strings`. 

However I did add f16 support using the [half](https://crates.io/crates/half) crate. It could also serve as an example where to change the code if a new type is added to od. 

The huge functions are mostly refactored to smaller functions. 

So I'll update the od status to:

 * [ ] od (almost complete, `--strings` and 128-bit datatypes are missing)

I used code like `something().unwrap_err();` to test for errors in unit tests. It is probably better to change that into `assert!(something().is_err());`, as it expresses the intent better. I suggest that I change that together with some other small changes in a cleanup commit.

Arcterus commented:
This is so much code, wow.  It's probably gonna take at least a few days to look over everything.

Arcterus commented on src/od/parse_formats.rs:
```diff
@@ -0,0 +1,525 @@
+                    expect_type_string = true;
+                }
+                else {
+                    match od_argument_traditional_format(c) {
```
This could be:
```rust
if let Some(r) = od_argument_traditional_format(c) {
    formats.push(ParsedFormatterItemInfo::new(r, false));
}
```

Arcterus commented on src/od/inputdecoder.rs:
```diff
@@ -0,0 +1,184 @@
+    /// `byte_order` determines how to read multibyte formats from the buffer.
+    pub fn new(input: &mut I, normal_length: usize, peek_length: usize, byte_order: ByteOrder) -> InputDecoder<I> {
+
+        let mut bytes: Vec<u8> = Vec::with_capacity(normal_length+peek_length);
```
Change to `normal_length + peek_length`.  Please make this correction throughout the code.

Arcterus commented on src/od/inputdecoder.rs:
```diff
@@ -0,0 +1,184 @@
+}
+
+
+impl<'a, I> InputDecoder<'a, I> where I : PeekRead {
```
`I: PeekRead`.  Also remove the extra space between variable names and colons throughout the code.

Arcterus commented on src/od/mockstream.rs:
```diff
@@ -0,0 +1,102 @@
+        if self.repeat_count == 0 {
+            return Ok(0)
+        }
+        else {
```
Please put the `else` on the same line as the `}`.  Make changes throughout the code.

Arcterus commented on src/od/multifilereader.rs:
```diff
@@ -0,0 +1,196 @@
+use std::vec::Vec;
+
+pub enum InputSource<'a> {
+    FileName(&'a str ),
```
Extra space.

Arcterus commented on src/od/multifilereader.rs:
```diff
@@ -0,0 +1,196 @@
+            any_err: false,
+        };
+        mf.next_file();
+        return mf;
```
`return` is not needed.

Arcterus commented on src/od/multifilereader.rs:
```diff
@@ -0,0 +1,196 @@
+        loop {
+            if self.ni.len() == 0 {
+                    self.curr_file = None;
+                    return;
```
I would personally prefer you use `break` in this situation.

Arcterus commented on src/od/multifilereader.rs:
```diff
@@ -0,0 +1,196 @@
+}
+
+impl<'b> io::Read for MultifileReader<'b> {
+
```
Extra newline.

Arcterus commented on src/od/od.rs:
```diff
@@ -46,446 +104,345 @@ pub fn uumain(args: Vec<String>) -> i32 {
+            Ok(CommandLineInputs::FileAndOffset((f, s, l))) => {
+                skip_bytes = s;
+                label = l;
+                vec!{f}
```
I don't see people use curly braces for `vec!` very often.  Usually they use square brackets or parentheses.

wimh commented:
Regarding all formatting issues; would it be useful to use [rustfmt](https://github.com/rust-lang-nursery/rustfmt) to fix those? In that case I would first add a commit adding `#[cfg_attr(rustfmt, rustfmt_skip)]` to functions which should not be formatted automatically.

Otherwise, if I manually fix those, should I create a separate commit containing only changed spacing (including empty lines), or can add it together with textual changes (like removing a redundant `return`)?

Arcterus commented:
You could try to use rustfmt and see how well it formats everything.  If that doesn't work out and you decide to manually format everything, you can add the changes together with textual ones.

wimh commented:
I chose not to use rustfmt, but tried to fix as much as possible with regular expressions, see below.
I also had to fix a build error on the beta and nightly channel.

```
sed -i 's/ :/:/' *.rs
sed -ri $'s/([^[\'"e ])\\+([^=) ])/\\1 + \\2/g' *.rs    
sed -ri $'s/([0-9])\\*([0-9])/\\1 * \\2/g' *.rs    
sed -i 's/\([^ ]\) )/\1)/' *.rs   
sed -i 's/vec!{\(.*\)}/vec![\1]/' *.rs 
perl -0777 -pi -e $'s/{\n\n/{\n/g' *.rs
perl -0777 -pi -e $'s/}\n([ ]*)else/} else/g' *.rs
perl -0777 -pi -e $'s/\n\n}/\n}/g' *.rs 
sed -i $'s/^\\([^"\']*[^ ]\\)==\\([^ ]\\)/\\1 == \\2/' *.rs  
sed -i $'s/^\\([^"\']*[^ ]\\)=\\([^"\' ]\\)/\\1 = \\2/' *.rs  
```

Arcterus commented on src/od/multifilereader.rs:
```diff
@@ -0,0 +1,194 @@
+        // loop retries with subsequent files if err - normally 'loops' once
+        loop {
+            if self.ni.len() == 0 {
+                    self.curr_file = None;
```
The indentation here is 8 instead of 4 for some reason.

wimh commented:
I fixed the indentation and squashed the last 5 (small) commits. I am hesitated to squash more as it makes reviews more difficult (at least I think so).

There were some intermittent broken pipe issues on travis:
```
---- test_od::test_invalid_offset stdout ----
	current_directory_resolved: 
run: /home/travis/build/uutils/coreutils/target/debug/uutils od -Ab
thread 'test_od::test_invalid_offset' panicked at 'Broken pipe (os error 32)', tests/common/util.rs:546
note: Run with `RUST_BACKTRACE=1` for a backtrace.
---- test_od::test_traditional_error stdout ----
	current_directory_resolved: 
run: /home/travis/build/uutils/coreutils/target/debug/uutils od --traditional 0 0 0 0
thread 'test_od::test_traditional_error' panicked at 'Broken pipe (os error 32)', tests/common/util.rs:546
```
I have been able to reproduce these after adding `sleep(Duration::from_secs(1));` on `tests/common/util.rs:538`. So the executable exits before the input can be piped into the process. 
These test verify error handling of the commandline arguments, so they don't read the input. I have removed the input in the unit tests and use `run` instead of `run_piped_stdin`. Afterwards that seems fixed.

Arcterus commented:
Again, thanks for this.  This is really a stunning amount of code. :)
